File: content/en/04-llms/llms/_exercises/10_structured-output/10_structured-output.py
# %%
import chatlas

# %%
from pyhere import here

recipe_txt = here("data/recipes/text/")
txt_cheesesteak = (recipe_txt / "PhillyCheesesteak.md").read_text()

# %%
print(txt_cheesesteak)

# %% [markdown]
# Here's an example of the structured output we want to achieve for a single
# recipe:
#
# ```json
# {
#   "title": "Spicy Mango Salsa Chicken",
#   "description": "A flavorful and vibrant chicken dish...",
#   "ingredients": [
#     {
#       "name": "Chicken Breast",
#       "quantity": "4",
#       "unit": "medium",
#       "notes": "Boneless, skinless"
#     },
#     {
#       "name": "Lime Juice",
#       "quantity": "2",
#       "unit": "tablespoons",
#       "notes": "Fresh"
#     }
#   ],
#   "instructions": [
#     "Preheat grill to medium-high heat.",
#     "In a bowl, combine ...",
#     "Season chicken breasts with salt and pepper.",
#     "Grill chicken breasts for 6-8 minutes per side, or until cooked through.",
#     "Serve chicken topped with the spicy mango salsa."
#   ]
# }
# ```
#
# Hint: include `None` in the type when the input might omit a value.

# %%
from pydantic import BaseModel, Field


class Ingredient(BaseModel):
    name: ____
    quantity: ____ | None = Field(description="____")
    unit: ____ | None = Field(description="____")
    notes: ____ | None = Field(description="____")


class Recipe(BaseModel):
    title: ____
    description: ____
    ingredients: list[_____]
    instructions: list[____] = Field(description="____")


# %%
chat = chatlas.ChatPosit(model="claude-sonnet-5")
recipe = chat.chat_structured(txt_cheesesteak, data_model=Recipe)

# %% [markdown]
# `.chat_structured()` returns an instance of the provided Pydantic model, so
# you can access fields directly:

# %%
recipe.title

# %% [markdown]
# Or you can convert it to JSON with pydantic's built-in `.model_dump_json()`
# method:

# %%
print(recipe.model_dump_json(indent=2))

File: content/en/04-llms/llms/_exercises/10_structured-output/10_structured-output.R
library(ellmer)

# Read in the recipes from text files
recipe_txt <- here::here("data/recipes/text")
txt_waffles <- recipe_txt |>
  file.path("CinnamonPeachOatWaffles.md") |>
  brio::read_file() # Like readLines() but all in one string

# Show the first 500 characters of the first recipe
txt_waffles |> substring(1, 500) |> cat()

#' Here's an example of the structured output we want to achieve for a single
#' recipe:
#'
#' {
#'   "title": "Spicy Mango Salsa Chicken",
#'   "description": "A flavorful and vibrant chicken dish...",
#'   "ingredients": [
#'     {
#'       "name": "Chicken Breast",
#'       "quantity": "4",
#'       "unit": "medium",
#'       "notes": "Boneless, skinless"
#'     },
#'     {
#'       "name": "Lime Juice",
#'       "quantity": "2",
#'       "unit": "tablespoons",
#'       "notes": "Fresh"
#'     }
#'   ],
#'   "instructions": [
#'     "Preheat grill to medium-high heat.",
#'     "In a bowl, combine ...",
#'     "Season chicken breasts with salt and pepper.",
#'     "Grill chicken breasts for 6-8 minutes per side, or until cooked through.",
#'     "Serve chicken topped with the spicy mango salsa."
#'   ]
#' }
#'
#' Hint: You can use `required = FALSE` in `type_*()` functions to indicate that
#' a field is optional.

type_recipe <- type_____(
  title = ____(),
  description = ____(),
  ingredients = ____(
    type_object(
      name = ____(),
      quantity = ____(),
      unit = ____(),
      notes = ____()
    )
  ),
  instructions = type_array(____())
)

chat <- chat_posit(model = "claude-sonnet-5")

chat$chat_structured(txt_waffles, type = type_recipe)
